src/Prove.hs

```diff
@@ -0,0 +1,249 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE DataKinds #-}
+
+module Prove (queries) where
+
+import Control.Monad (when)
+import Data.ByteString (ByteString)
+import Data.Either
+import Data.List
+import Data.List.NonEmpty as NonEmpty (NonEmpty, toList)
+import Data.Map.Strict as Map (Map, lookup, fromList, toList, unionWith, empty, keys)
+import Data.Maybe
+
+import Data.SBV
+import Data.SBV.Trans.Control
+
+import RefinedAst
+import Syntax (Id, Interface)
+import Type (metaType)
+
+-- *** Interface *** --
+
+
+{-|
+   For each Invariant claim build an SMT query consisting of:
+
+   - constants for the pre and post versions of all storage variables used by the transition system
+   - boolean predicates over the pre and post storage variables for each pass behaviour
+   - a boolean predicate for the invariant
+   - an assertion that the invariant does not hold if either of the following is true:
+      1. The constructor predicate holds
+      2. The invariant holds over the prestate and one of the method level predicates holds
+
+   If this query returns `unsat` then the invariant must hold over the transition system
+-}
+queries :: [Claim] -> [Symbolic ()]
+queries claims = fmap mkQuery $ gather claims
+
+
+-- *** Data *** --
+
+
+data When = Pre | Post
+  deriving (Show)
+
+data Ctx = Ctx Id Id When [(Id, SMType)]
+  deriving (Show)
+
+data SMType
+  = SymInteger (SBV Integer)
+  | SymBool (SBV Bool)
+  | SymBytes (SBV [WordN 8])
+  deriving (Show)
+
+catInts :: [(Id, SMType)] -> [(Id, SBV Integer)]
```

Comment:
another alternative would be to use `Data.Bifunctor` to do something along the lines of:
```
catInts = fmap $ second (\(SymInteger i) -> i)
```